---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# conservedPos

<!-- badges: start -->
<!-- badges: end -->

## Description
Description: The goal of conservedPos package is to find important conserved sites for small target region by conducting a site-specific analysis for a set of aligned DNA sequences. 
The package will provide function to count nucleotides occurrances in each position independently and calculates the frequency of the nucleotides in order to quantify the conservity of each position.

## Installation

You can install the released version of conservedPos 

``` r
install.packages("conservedPos")

# install.packages("devtools")
library(devtools)
require

devtools::install_github("hezijin/conservedPos", build_vignettes = TRUE)
```
## Example

## Acknowledgements

This package was developed as part of an assessment for 2020BCB410H: Applied Bioinformatics, University of Toronto, Toronto,CANADA.
